bookkeeper-server/src/main/java/org/apache/bookkeeper/client/UpdateLedgerOp.java

```diff
@@ -81,174 +87,80 @@ public UpdateLedgerOp(final BookKeeper bkc, final BookKeeperAdmin admin) {
-                    final CountDownLatch syncObj = new CountDownLatch(1);
-
-                    // iterate through all the ledgers
-                    while (ledgerItr.hasNext() && !stop.get()) {
-                        // throttler to control updates per second
-                        throttler.acquire();
-
-                        final Long lId = ledgerItr.next();
-                        final ReadLedgerMetadataCb readCb = new ReadLedgerMetadataCb(bkc, lId, oldBookieId,
-                                newBookieId);
-                        outstandings.add(lId);
-
-                        FutureCallback<Void> updateLedgerCb = new UpdateLedgerCb(lId, stop, issuedLedgerCnt,
-                                updatedLedgerCnt, outstandings, syncObj, progressable);
-                        Futures.addCallback(readCb.getFutureListener(), updateLedgerCb);
-
-                        issuedLedgerCnt.incrementAndGet();
-                        if (limit != Integer.MIN_VALUE && issuedLedgerCnt.get() >= limit || !ledgerItr.hasNext()) {
-                            stop.set(true);
+        final AtomicBoolean errorOccurred = new AtomicBoolean(false);
+        final Set<CompletableFuture<?>> outstanding =
+            Collections.newSetFromMap(new ConcurrentHashMap<CompletableFuture<?>, Boolean>());
+        final RateLimiter throttler = RateLimiter.create(rate);
+        final Iterator<Long> ledgerItr = admin.listLedgers().iterator();
+
+        // iterate through all the ledgers
+        while (ledgerItr.hasNext() && !errorOccurred.get()
+               && (limit == Integer.MIN_VALUE || issuedLedgerCnt.get() < limit)) {
+            // throttler to control updates per second
+            throttler.acquire();
+
+            final long ledgerId = ledgerItr.next();
+            issuedLedgerCnt.incrementAndGet();
+
+            GenericCallbackFuture<LedgerMetadata> promise = new GenericCallbackFuture<>();
+            lm.readLedgerMetadata(ledgerId, promise);
+            promise.thenCompose((readMetadata) -> {
+                    AtomicReference<LedgerMetadata> ref = new AtomicReference<>(readMetadata);
+                    return new MetadataUpdateLoop(
+                            lm, ledgerId,
+                            ref::get,
+                            (metadata) -> {
+                                return metadata.getEnsembles().values().stream()
+                                    .flatMap(Collection::stream)
+                                    .filter(b -> b.equals(oldBookieId))
+                                    .count() > 0;
+                            },
+                            (metadata) -> {
+                                return replaceBookieInEnsembles(metadata, oldBookieId, newBookieId);
+                            },
+                            ref::compareAndSet).run();
+                }).whenComplete((metadata, ex) -> {
+                        LOG.info("Update of {} finished", ledgerId);
```

Comment:
`update` is ambiguous. can you make the statement more specific?

Reply:
done.